File: WordPressAuthenticator/Services/WordPressComOAuthClientFacade.m
#import "WordPressComOAuthClientFacade.h"
#import "WPAuthenticator-Swift.h"
@import WordPressKit;

@interface WordPressComOAuthClientFacade ()

@property (nonatomic, strong) WordPressComOAuthClient *client;

@end

@implementation WordPressComOAuthClientFacade

@synthesize client;

- (instancetype)initWithClient:(NSString *)client secret:(NSString *)secret
{
    NSParameterAssert(client);
    NSParameterAssert(secret);
    self = [super init];
    if (self) {
        self.client = [WordPressComOAuthClientFacade initializeOAuthClientWithClientID:client secret:secret];
    }

    return self;
}

- (instancetype)init {
    NSAssert(false, @"Please initializer WordPressComOAuthClientFacade with the ClientID and Secret!");
    return nil;
}

- (void)authenticateWithUsername:(NSString *)username
                        password:(NSString *)password
                 multifactorCode:(NSString *)multifactorCode
                         success:(void (^)(NSString *authToken))success
                needsMultiFactor:(void (^)(NSInteger userID, SocialLogin2FANonceInfo *nonceInfo))needsMultifactor
                         failure:(void (^)(NSError *error))failure
{
    [self.client authenticateWithUsername:username password:password multifactorCode:multifactorCode needsMultifactor:needsMultifactor success:success failure:^(NSError * error) {
        if (error.code == WordPressComOAuthErrorNeedsMultifactorCode) {
            if (needsMultifactor != nil) {
                needsMultifactor(0, nil);
            }
        } else {
            if (failure != nil) {
                failure(error);
            }
        }
    }];
}

- (void)requestOneTimeCodeWithUsername:(NSString *)username
                              password:(NSString *)password
                               success:(void (^)(void))success
                               failure:(void (^)(NSError *error))failure
{
    [self.client requestOneTimeCodeWithUsername:username password:password success:success failure:failure];
}

- (void)requestSocial2FACodeWithUserID:(NSInteger)userID
                                 nonce:(NSString *)nonce
                               success:(void (^)(NSString *newNonce))success
                               failure:(void (^)(NSError *error, NSString *newNonce))failure
{
    [self.client requestSocial2FACodeWithUserID:userID nonce:nonce success:success failure:failure];
}

- (void)authenticateWithSocialIDToken:(NSString *)token
                              service:(NSString *)service
                              success:(void (^)(NSString *authToken))success
                     needsMultiFactor:(void (^)(NSInteger userID, SocialLogin2FANonceInfo *nonceInfo))needsMultifactor
          existingUserNeedsConnection:(void (^)(NSString *email))existingUserNeedsConnection
                              failure:(void (^)(NSError *error))failure
{
    [self.client authenticateWithIDToken:token
                                 service:service
                                 success:success
                        needsMultifactor:needsMultifactor
             existingUserNeedsConnection:existingUserNeedsConnection
                                 failure:failure];
}

- (void)authenticateSocialLoginUser:(NSInteger)userID
                           authType:(NSString *)authType
                        twoStepCode:(NSString *)twoStepCode
                       twoStepNonce:(NSString *)twoStepNonce
                            success:(void (^)(NSString *authToken))success
                            failure:(void (^)(NSError *error))failure
{
    [self.client authenticateSocialLoginUser:userID authType:authType twoStepCode:twoStepCode twoStepNonce:twoStepNonce success:success failure:failure];
}

- (void) requestWebauthnChallengeWithUserID: (NSInteger)userID
                               twoStepNonce:(NSString *)twoStepNonce
                                    success:(void (^)(WebauthnChallengeInfo *challengeData))success
                                    failure:(void (^)(NSError *error))failure {
    [self.client requestWebauthnChallengeWithUserID:userID twoStepNonce:twoStepNonce success:success failure:failure];
}

- (void) authenticateWebauthnSignatureWithUserID:(NSInteger)userID
                                    twoStepNonce:(NSString *)twoStepNonce
                                    credentialID:(NSData *)credentialID
                                  clientDataJson:(NSData *)clientDataJson
                               authenticatorData:(NSData *)authenticatorData
                                       signature:(NSData *)signature
                                      userHandle:(NSData *)userHandle
                                         success:(void (^)(NSString *authToken))success
                                         failure:(void (^)(NSError *error))failure {
    [self.client authenticateWebauthnSignatureWithUserID:userID
                                            twoStepNonce:twoStepNonce
                                            credentialID:credentialID
                                          clientDataJson:clientDataJson
                                       authenticatorData:authenticatorData
                                               signature:signature
                                              userHandle:userHandle
                                                 success:success
                                                 failure:failure];
}

@end

File: WordPressAuthenticator/Services/WordPressComOAuthClientFacade.swift
import Foundation
import WordPressKit

@objc public class WordPressComOAuthClientFacade: NSObject, WordPressComOAuthClientFacadeProtocol {

    private let client: WordPressComOAuthClient

    @objc required public init(client: String, secret: String) {
        self.client = WordPressComOAuthClient(clientID: client,
                                         secret: secret,
                                         wordPressComBaseUrl: WordPressAuthenticator.shared.configuration.wpcomBaseURL,
                                         wordPressComApiBaseUrl: WordPressAuthenticator.shared.configuration.wpcomAPIBaseURL)
    }

    public func authenticate(
        username: String,
        password: String,
        multifactorCode: String?,
        success: @escaping (_ authToken: String?) -> Void,
        needsMultifactor: ((_ userID: Int, _ nonceInfo: SocialLogin2FANonceInfo?) -> Void)?,
        failure: ((_ error: Error) -> Void)?
    ) {
        self.client.authenticateWithUsername(username, password: password, multifactorCode: multifactorCode, needsMultifactor: needsMultifactor, success: success, failure: { error in
            if error.code == WordPressComOAuthError.needsMultifactorCode.rawValue {
                needsMultifactor?(0, nil)
            } else {
                failure?(error)
            }
        })
    }

    public func requestOneTimeCode(
        username: String,
        password: String,
        success: @escaping () -> Void,
        failure: @escaping (_ error: Error) -> Void
    ) {
        self.client.requestOneTimeCodeWithUsername(username, password: password, success: success, failure: failure)
    }

    public func requestSocial2FACode(
        userID: Int,
        nonce: String,
        success: @escaping (_ newNonce: String) -> Void,
        failure: @escaping (_ error: Error, _ newNonce: String?) -> Void
    ) {
        self.client.requestSocial2FACodeWithUserID(userID, nonce: nonce, success: success, failure: failure)
    }

    public func authenticate(
        socialIDToken: String,
        service: String,
        success: @escaping (_ authToken: String?) -> Void,
        needsMultifactor: @escaping (_ userID: Int, _ nonceInfo: SocialLogin2FANonceInfo) -> Void,
        existingUserNeedsConnection: @escaping (_ email: String) -> Void,
        failure: @escaping (_ error: Error) -> Void
    ) {
        self.client.authenticateWithIDToken(
            socialIDToken,
            service: service,
            success: success,
            needsMultifactor: needsMultifactor,
            existingUserNeedsConnection: existingUserNeedsConnection,
            failure: failure
        )
    }

    public func authenticate(
        socialLoginUser userID: Int,
        authType: String,
        twoStepCode: String,
        twoStepNonce: String,
        success: @escaping (_ authToken: String?) -> Void,
        failure: @escaping (_ error: Error) -> Void
    ) {
        self.client.authenticateSocialLoginUser(
            userID,
            authType: authType,
            twoStepCode: twoStepCode,
            twoStepNonce: twoStepNonce,
            success: success,
            failure: failure
        )
    }

    public func requestWebauthnChallenge(
        userID: Int64,
        twoStepNonce: String,
        success: @escaping (_ challengeData: WebauthnChallengeInfo) -> Void,
        failure: @escaping (_ error: Error) -> Void
    ) {
        self.client.requestWebauthnChallenge(userID: userID, twoStepNonce: twoStepNonce, success: success, failure: failure)
    }

    public func authenticateWebauthnSignature(
        userID: Int64,
        twoStepNonce: String,
        credentialID: Data,
        clientDataJson: Data,
        authenticatorData: Data,
        signature: Data,
        userHandle: Data,
        success: @escaping (_ authToken: String) -> Void,
        failure: @escaping (_ error: Error) -> Void
    ) {
        self.client.authenticateWebauthnSignature(
            userID: userID,
            twoStepNonce: twoStepNonce,
            credentialID: credentialID,
            clientDataJson: clientDataJson,
            authenticatorData: authenticatorData,
            signature: signature,
            userHandle: userHandle,
            success: success,
            failure: failure
        )
    }

}
